errbot/core_plugins/webserver.py

```diff
@@ -74,48 +76,61 @@ def get_configuration_template(self):
                         'host': '0.0.0.0',
                         'port': 3142,
-                        'certificate': "",
-                        'key': ""}}
+                        'certificate': '',
+                        'key': ''}}
 
     def check_configuration(self, configuration):
         # it is a pain, just assume a default config if SSL is absent or set to None
         if configuration.get('SSL', None) is None:
-            configuration['SSL'] = {'enabled': False, 'host': '0.0.0.0', 'port': 3142, 'certificate': "", 'key': ""}
+            configuration['SSL'] = {'enabled': False, 'host': '0.0.0.0', 'port': 3142, 'certificate': '', 'key': ''}
         super().check_configuration(configuration)
 
     def activate(self):
         if not self.config:
             self.log.info('Webserver is not configured. Forbid activation')
             return
 
-        host = self.config['HOST']
-        port = self.config['PORT']
-        ssl = self.config['SSL']
-        interfaces = [(host, port)]
-        if ssl['enabled']:
-            # noinspection PyTypeChecker
-            interfaces.append((ssl['host'], ssl['port'], ssl['key'], ssl['certificate']))
-        self.log.info('Firing up the Rocket')
-        self.webserver = Rocket(interfaces=interfaces,
-                                app_info={'wsgi_app': bottle_app}, )
-        self.webserver.start(background=True)
-        self.log.debug('Liftoff!')
+        if self.server_thread and self.server_thread.is_alive():
+            raise Exception('Invalid state, you should not have a webserver already running.')
+        self.server_thread = Thread(target=self.run_server, name='Webserver Thread')
+        self.server_thread.start()
+        self.log.debug('Webserver started.')
 
         super().activate()
 
     def deactivate(self):
-        if self.webserver is not None:
-            self.log.debug('Sending signal to stop the webserver')
-            self.webserver.stop()
+        if self.server is not None:
+            self.log.info('Shutting down the internal webserver.')
+            self.server.shutdown()
+            self.log.info('Waiting for the webserver thread to quit.')
+            self.server_thread.join()
+            self.log.info('Webserver shut down correcly.')
         super().deactivate()
 
-    # noinspection PyUnusedLocal
+    def run_server(self):
+        try:
+            host = self.config['HOST']
+            port = self.config['PORT']
+            ssl = self.config['SSL']
+            self.log.info('Starting the webserver on %s:%i' % (host, port))
+            ssl_context = (ssl['certificate'], ssl['key']) if ssl['enabled'] else None
+            self.server = ThreadedWSGIServer(host, ssl['port'] if ssl_context else port, flask_app,
+                                             ssl_context=ssl_context)
+            self.server.serve_forever()
+            self.log.debug('Webserver stopped')
+        except KeyboardInterrupt as _:
```

Comment:
Maybe do `except KeyboardInterrupt:` (omitting the `as _` bit) since the exception object is not used?